### DISCUSSION TITLE
balancer phase for custom plugin for restricting upstream IP address

### DISCUSSION THREAD

**Discussion by rucciva** [Kong/kong#7322] [Ideas and feature requests]
### Summary

In our deployment, kong is used to expose multiple APIs from multiple teams. Each teams can use hostname or ip as the upstream information. Since hostname is externally controlled, it is possible for a team to change the IP of the hostname and might use it to expose sensitive backend accessible by kong, such as but not limited to kong's admin APIs located at 127.0.0.1. Thus we need a way--preferably using custom plugin--to restrict upstream IP address. 

I've looked at the kong source code and (CMIIW) found out that upstream ip address are only resolved after access phase and available at balancer phase. So maybe make balancer phase available for custom plugin that is invoked in Kong.balancer() ?

Thanks.

**Comment by bungle** [comment]
We actually resolve ip before balancer phase (on access phase), but the plugins have already ran the access phase at that point. We hope to make it plugins more flexible. At the moment it is not possible to do what you ask as a plugin, but thank you for suggestion.

**Reply by AshayFernandes** [reply]
We were also building a custom kong plugin, which required destination IP address in access phase(before proxying the request) that the kong resolved to, using the Host(an srv record) field of service entity. Are there any plans of providing this functionality?
